### PR TITLE
Cache channel offline banners and game box art

### PR DESCRIPTION
**Pull request by dchristidis**
Banners and box art change rarely. By caching them to disk, the start-up time of GNOME Twitch can be substantially improved. When I tested this, with 10 favorite channels and a hot cache, the amount of data transferred on start-up dropped from about 12 MB to just over 2 MB.

The last commit as a bit of a gimmick, but what it means is that cached images will be considered fresh for 7 to 14 days. I think that this is a reasonable lifespan.

### CODE REVIEW

**Comment by dengelt** [comment]
As an alternative to waiting a set time before downloading the files again, the app could send HTTP HEAD requests in the background after the cached images have been loaded, and re-download and replace them if they have been modified in the meantime. This can be checked by comparing the `Last-Modified` field in the HTTP header to the last modification date of the cached file.

A HTTP header is rather small, so it wouldn't use too much bandwidth, and the startup time should be unaffected since the files are loaded from cache first and the user can already use the app with the cached files while they are checked and possibly updated.

**Comment by Ippytraxx** [comment]
The code looks good but I think @dengelt solution is better. Just tried it out and sending a HEAD request works fine. There is also a Expires field that can be used.

On top of that other normal API requests set an Etag that might be useful in the future but most API data would change quite frequently so I don't know if it's worth caching that data.

**Comment by dchristidis** [comment]
Indeed, that is superior. Plus, it gets rid of that randomness gimmick. I'll work on it.

**Comment by dchristidis** [comment]
I'm making some good progress. In the meantime, could this be implemented using the caching capabilities provided by libsoup? I tried it, but I couldn't get it to work.

**Comment by dengelt** [comment]
Do you know how libsoup handles caching? While looking through the possible HTTP header fields, I stumbled upon `If-Modified-Since`, which could simplify caching a bit (and save a round-trip time and a few bytes of bandwidth if the cache is stale). However, when I tested if the twitch servers handled this field correctly, they didn't. Or I did something wrong. I used curl for testing. Relevant parts of the request:

```
curl -v -H "If-Modified-Since: Wed, 06 Apr 2016 15:18:13 GMT" https://static-cdn.jtvnw.net/jtv_user_pictures/ptken-profile_banner-299e0406dc15485b-480.jpeg > test.jpeg

[...]
> GET /jtv_user_pictures/ptken-profile_banner-299e0406dc15485b-480.jpeg HTTP/1.1
> Host: static-cdn.jtvnw.net
> User-Agent: curl/7.47.0
> Accept: */*
> If-Modified-Since: Wed, 06 Apr 2016 15:18:13 GMT
> 
< HTTP/1.1 200 OK
< Date: Wed, 06 Apr 2016 15:19:49 GMT
< Content-Type: image/jpeg
< Content-Length: 145798
< Connection: keep-alive
< ETag: "4d0233aa8ea89781c2356284bd5033ef"
< Last-Modified: Wed, 23 Dec 2015 17:20:41 GMT
< Server: nginx
< x-amz-id-2: nYtuDN0XO6n34PwZJipP1sTGlDu4rSt1Cp4kjka1PikUvOfOOxl048lG4OzT1jvdbsXHsMTitOM=
< x-amz-request-id: 72699721D325F549
< Accept-Ranges: bytes
< 
[...]
```

In theory, the server should respond with a `304 Not Modified` if the last modification time was before the specified time, but the server still sent the file. So if libsoup uses the same mechanism, maybe that's the problem. I didn't try to use the ETag field yet, because the twitch servers don't seem to include it for all files (they seem to be missing for offline video banners for example).

**Comment by Ippytraxx** [comment]
@dengelt That is most likely what libsoup is trying to do (https://github.com/GNOME/libsoup/blob/master/tests/cache-test.c). Unfortunately the Twitch API is horrendously broken so we have to do the cache validation on our side. Basically just saving the files with a corresponding JSON list with filenames and dates, then sending HTTP HEAD requests to check dates. It seems to only set ETags on text responses but not images.

**Comment by dengelt** [comment]
The image I used for testing in the comment above came with an ETag field, but others (like offline video banners) didn't. After testing with this image, I found that Twitch indeed answers with a `304 Not Modified` if I supply a matching ETag via `If-None-Match`:

```
curl -v -H "If-None-Match: \"4d0233aa8ea89781c2356284bd5033ef\"" https://static-cdn.jtvnw.net/jtv_user_pictures/ptken-profile_banner-299e0406dc15485b-480.jpeg > test.jpeg

[...]
> GET /jtv_user_pictures/ptken-profile_banner-299e0406dc15485b-480.jpeg HTTP/1.1
> Host: static-cdn.jtvnw.net
> User-Agent: curl/7.47.0
> Accept: */*
> If-None-Match: "4d0233aa8ea89781c2356284bd5033ef"
> 
< HTTP/1.1 304 Not Modified
< Date: Wed, 06 Apr 2016 16:12:41 GMT
< Connection: keep-alive
< ETag: "4d0233aa8ea89781c2356284bd5033ef"
< Last-Modified: Wed, 23 Dec 2015 17:20:41 GMT
< Server: nginx
< x-amz-id-2: ar1wwaMkASWEn5bZxG61LZIwczwAKPomvfi+4wnrssDbwPNlYH3iJPdQAARqveCHNQa8n2lzO84=
< x-amz-request-id: E11023B13315D937
< 
[...]
```

Interestingly though, when I used both `If-None-Match` and `If-Modified-Since` in the same request, Twitch sent the image again. So the easiest way would probably be to just use HEAD requests like I originally proposed.

**Comment by dchristidis** [comment]
I would like to make a request for comments on the current progress.

If there are no objections, I will use the same logic for channel banners. An added bonus is that HEAD requests can be used for live channel previews too.

File: src/gt-channel.c [hunk 1]
```diff
@@ -218,14 +222,33 @@ update_preview(GtChannel* self)
...
     }
     else
     {
         if (priv->video_banner_url)
-            gt_twitch_download_picture_async(main_app->twitch, priv->video_banner_url, priv->cancel, 
-                                             (GAsyncReadyCallback) download_picture_cb, self); 
+        {
+            if (!g_file_test(priv->cache_filename, G_FILE_TEST_EXISTS))
+                g_info("{GtChannel} Cache miss for channel '%s'", priv->name);
+            else if (utils_timestamp_now() - utils_timestamp_file(priv->cache_filename) > 604800 + g_random_int_range(0, 604800))
```
Comment (dengelt) [inline src/gt-channel.c]: Shouldn't this check be removed, since the "staleness" of the cache file is determined by the HEAD request now?

Comment (dchristidis) [inline src/gt-channel.c]: You are correct. However, this is not expected to be the final state of the code. At this time, the caching policy that was originally proposed is still in use for channel banners. The aim, of course, is to implement your suggestion.

I'm looking for feedback mainly on `gt-twitch.c`. Afterwards, I intend to apply the same logic to `gt-channel.c`

**Comment by Ippytraxx** [comment]
Seems good so far. The only thing is that when you are changing GtChannel make sure you keep it so that the spinner for each one is visible while fetching the images (like it is now). I was planning on making GtGame do the same but I'll wait until you are done.

Basically just so that there aren't a whole bunch of empty boxes that suddenly get populated with images.

**Comment by dchristidis** [comment]
I'm not sure I understand.

When `gt_twitch_top_games()` is called, if there's a cache miss, the picture will be downloaded immediately. If there's a cache hit, the picture is loaded from disk. Thus, when the function call returns, all  games will have a picture.

The update happens in the background, as @dengelt suggested. If the picture has been modified, it's downloaded and it replaces the old one. No form of animation is used, but this won't be a problem. Box art seems to change only upon a new release. This update process is not expected to happen often.

The same model will be used for channel banners.

Were you planning to change the games view to first display the name of the games and then load their pictures asynchronously? With caching, the games load _really_ fast. Wouldn't it be friendlier to the eye if the user waits a couple of seconds and all games appear at the same time, than to have multiple images pop-in all the time?

It's not like the channels view. The pictures are much smaller and the cache has great performance. Loading, scrolling through and even searching can take less than a second.

**Comment by Ippytraxx** [comment]
Hmm, I've only looked at the code online so I might've missed something but I see you using a thread pool to download pictures. Then in the callback you use `g_object_set` to set the image, doesn't this cause `GtTwitchTopGames` to return before it has downloaded all the pictures?

**Comment by dchristidis** [comment]
Thanks for taking the time to provide feedback on this pull request. It turned out to be quite extensive and I hoped that a small pause would make review process easier.

I've tested it as much as I could, but I'm still a bit concerned about thread safety and memory management.

**Comment by Ippytraxx** [comment]
Works well! 

I agree on the thread safety and memory management, that's why I'll probably move all the thread pools and stuff to a centralised place like GtChannelsManager so it's easier to keep track of.

**Pull request merged**